Collections Completeness Tests

john-westcott-iv opened the pull request:
##### SUMMARY
Adds a test to test_collection which validates for feature completness.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - Collection

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 14.0.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

john-westcott-iv committed:
Initial commit of test_completeness

AlanCoding commented:
This looks really awesome.

Getting into the technical nitty gritty, I'm interested if we could "map" `tower_workflow_launch` to `workflow_jobs`, because the field comparisons should still be meaningful there. The editable / read-only distinction won't be accurate, but I objectively we would still like to look for field parity there.

john-westcott-iv committed:
Adding read-only endpoint check, fixing typo, adding needs_development check

john-westcott-iv committed:
Spell checking

john-westcott-iv committed:
Fixing pep8 issues

john-westcott-iv committed:
Changing job_timeout to an aliase of timeout to match the API

john-westcott-iv committed:
Changing scm_credential to credential to match API

john-westcott-iv committed:
Fixing flake8 and adding more module OKs

john-westcott-iv committed:
Change module from tower_notification to tower_notification_template

john-westcott-iv commented:
Adding reference to #7353

john-westcott-iv committed:
Migrating old test_notification test to test_notification_template

john-westcott-iv committed:
Fixing pep8 and other issues

john-westcott-iv committed:
Fixing PY2 compile issues

AlanCoding commented on awx_collection/test/awx/test_completeness.py:
```diff
@@ -0,0 +1,251 @@
+# Noramlly a read-only endpoint should not have a module (i.e. /api/v2/me) but sometimes we reuse a name
+# For example, we have a tower_role module but /api/v2/roles is a read only endpoint.
+# Rhis list indicates which read-only endpoints have associated modules with them.
+read_only_endpoints_with_modules = ['tower_settings', 'tower_role']
```
But both of these wind up being exceptions anyway don't they?

AlanCoding commented on awx_collection/test/awx/test_completeness.py:
```diff
@@ -0,0 +1,251 @@
+    'tower_credential': [
+        'authorize', 'authorize_password', 'become_method', 'become_password', 'become_username', 'client',
+        'domain', 'host', 'kind', 'password', 'project', 'secret', 'security_token', 'ssh_key_data',
+        'ssh_key_unlock', 'subscription', 'tenant', 'username', 'vault_id', 'vault_password',
```
instead of enumerating these, could you instead search for "deprecated" in the help text and skip in those cases instead?

john-westcott-iv replied:
Deprecated might catch some of them but not sure I want to rely specifically on that.

AlanCoding commented on awx_collection/test/awx/test_completeness.py:
```diff
@@ -0,0 +1,251 @@
+    'tower_notification_template': [
+        'account_sid', 'account_token', 'channels', 'client_name', 'color', 'from_number', 'headers', 'host',
+        'message_from', 'nickname', 'notify', 'password', 'port', 'recipients', 'sender', 'server',
+        'service_key', 'subdomain', 'targets', 'to_numbers', 'token', 'url', 'use_ssl', 'use_tls', 'username',
```
Same here, I would prefer some more general solution for deprecated params

AlanCoding commented on awx_collection/test/awx/test_completeness.py:
```diff
@@ -0,0 +1,251 @@
+    'tower_job_template': ['credential', 'vault_credential', 'credentials', 'labels', 'survey_spec', 'organization'],
+    # Always_nodes, failure_nodes, success_nodes, credentials is how we handle associations
+    # Organization is how we looking job templates
+    'tower_workflow_job_template_node': ['always_nodes', 'failure_nodes', 'success_nodes', 'credentials', 'organization'],
```
Could we also do some general solution (via parameter help text) for related fields?

People are general averse to hard lists of fields, and in cases like these, we can get real value added for the user by putting this signals in a standard form in the help text in DOCUMENTATION.

AlanCoding commented on awx_collection/test/awx/test_completeness.py:
```diff
@@ -0,0 +1,251 @@
+# that needs to be developed. If the module is found on the file system it will auto-detect that the
+# work is being done and will bypass this check. At some point this module should be removed from this list.
+needs_development = [
+    'tower_ad_hoc_command', 'tower_application', 'tower_instance_group', 'tower_inventory_script',
```
`tower_inventory_script` is for a deprecated endpoint, so you could move this over to the ignore list instead.

john-westcott-iv committed:
Changing how we check modules

Options which are not in the API POST and are marked in the module as deprecated are ignored

If an option is not in the API POST but is marked as a list we assume its a relation

john-westcott-iv committed:
Permissions change, spell check, pep8, etc

AlanCoding reviewed (approved):
Thanks for implementing the changes from our discussion, the exceptions list seems very reasonable to me here. I'm good to merge this as something that closes #7353
